src/Collection/ArrayCollection.php

```diff
@@ -0,0 +1,260 @@
+        return $new;
+    }
+
+    /**
+     * @param array|self ...$args
+     * @return self
+     */
+    public function mergeWith(...$args): self
+    {
+        array_unshift($args, $this);
+
+        $arrays = [];
+        foreach ($args as $arg) {
+            $arrays[] = $arg instanceof self ? $arg->data : $arg;
+        }
+
+        $collections = [];
+        foreach ($args as $index => $arg) {
+            $collection = $arg instanceof self ? $arg : new self($arg);
+            foreach ($collection->getModifiers() as $modifier) {
+                if ($modifier instanceof BeforeMergeModifierInterface) {
+                    $collection->data = $modifier->beforeMerge($arrays, $index);
+                }
+            }
+            $collections[$index] = $collection;
+        }
+
+        $collection = $this->merge(...$collections);
+
+        foreach ($collection->getModifiers() as $modifier) {
+            if ($modifier instanceof AfterMergeModifierInterface) {
+                $collection->data = $modifier->afterMerge($collection->data);
+            }
+        }
+
+        return $collection;
+    }
+
+    /**
+     * @param array|self ...$args
+     * @return self
+     */
+    private function merge(...$args): self
+    {
+        $collection = new ArrayCollection();
+
+        while (!empty($args)) {
+            $array = array_shift($args);
+
+            if ($array instanceof ArrayCollection) {
+                $collection->modifiers = array_merge($collection->modifiers, $array->modifiers);
+                $collection->data = $this->merge($collection->data, $array->data)->data;
+                continue;
+            }
+
+            foreach ($array as $k => $v) {
+                if (is_int($k)) {
+                    if (array_key_exists($k, $collection->data)) {
+                        if ($collection->data[$k] !== $v) {
+                            $collection->data[] = $v;
+                        }
+                    } else {
+                        $collection->data[$k] = $v;
+                    }
+                } elseif (
+                    static::isMergable($v) &&
+                    isset($collection->data[$k]) &&
+                    static::isMergable($collection->data[$k])
+                ) {
+                    $mergedCollection = $this->merge($collection->data[$k], $v);
+                    $collection->data[$k] = ($collection->data[$k] instanceof self || $v instanceof self)
+                        ? $mergedCollection
+                        : $mergedCollection->data;
+                } else {
+                    $collection->data[$k] = $v;
+                }
+            }
+        }
+
+        return $collection;
+    }
+
+    /**
+     * @param mixed $value
+     * @return bool
+     */
+    private function isMergable($value): bool
+    {
+        return is_array($value) || $value instanceof self;
+    }
+
+    public function toArray(): array
+    {
+        if ($this->array === null) {
```

Comment:
I think it may result in stuck data since `array` is cloned along with collection:

```php
$arrayCollection = new ArrayCollection(['a' => 13, 'b' => 42]);
var_dump($arrayCollection->toArray());
$newCollection = $arrayCollection->withModifier(new UnsetValue('a'));
var_dump($newCollection->toArray()); // should give you ['b' => 42] but...
```